internal/website/content/howto/pubsub/subscribe.md

```diff
@@ -0,0 +1,504 @@
+---
+title: "Subscribe to a Topic's Messages"
+date: 2019-03-26T09:44:33-07:00
+draft: true
+weight: 2
+---
+
+Subscribing to messages on a topic with the Go CDK takes two steps:
+
+1. Open the subscription with the Pub/Sub provider of your choice (once per subscription).
+2. Receive messages from the topic.
+
+The second step is the same across all providers because the first step
+creates a value of the portable [`*pubsub.Subscription`][] type. A simple
+subscriber that operates on messages serially looks like this:
+
+```go
+func main() {
+    // Parse configuration and open subscription (shutdown deferred).
+    // ...
+
+    // ... and then loop on received messages.
+    for {
+        msg, err := subscription.Receive(ctx)
+        if err != nil {
+            // Errors from Receive indicate that Receive will no longer succeed.
+            log.Printf("Receiving message: %v", err)
+            break
+        }
+        // TODO(light): Is this reasonable ack behavior, @jba?
```

Comment:
Maybe. In the error case, was the error inherent to the message (e.g. it was invalid)? If so, "handle" it by logging, emailing someone, alerting, etc. Then ack it—you handled it, no one else needs to see it. If the error was transient (e.g. couldn't reach a backend), then it's okay to drop it, or even Nack it, so the message will be sent elsewhere.

Acking a message means no one needs to look at it again, ever. That cross-cuts the question of whether processing the message "fails". That is true for all at-least-once systems, so maybe refer people to the wikipedia page before they start this section, so they understand that. (Though it wouldn't hurt to lightly repeat it here.)

Reply:
> Maybe. In the error case, was the error inherent to the message (e.g. it was invalid)? If so, "handle" it by logging, emailing someone, alerting, etc. Then ack it—you handled it, no one else needs to see it. If the error was transient (e.g. couldn't reach a backend), then it's okay to drop it, or even Nack it, so the message will be sent elsewhere.

Makes sense. There isn't really a Go standard convention around what constitutes a transient error. The [`Temporary()` method](https://godoc.org/net#Error.Temporary) is probably about as good as it gets, but it's inconvenient to test for and isn't a surefire test, so perhaps prose documentation is good enough, but this does seem difficult for users in general and I'd like to put forward a good pattern.

> Acking a message means no one needs to look at it again, ever. That cross-cuts the question of whether processing the message "fails". That is true for all at-least-once systems, so maybe refer people to the wikipedia page before they start this section, so they understand that. (Though it wouldn't hurt to lightly repeat it here.)

I don't understand this. Can you provide a link to which Wikipedia page you mean, so I can understand it? 🙂

Reply:
It seems I was mistaken about the existence of that wikipedia page. @vangent, wasn't there a page that described this well?

Here's the idea in a nutshell. When you publish a message to an at-least-once system, it will keep trying to deliver it until it's acked. If it sends it to subscriber 1 and that subscribers "nacks" it (says "no, I can't accept this message"), then the system delivers it to someone else. If that subscriber doesn't ack it (before some predetermined deadline), the system will try again. It will keep trying (within limits) until it gets an ack. Conversely, once it gets an ack, it will never send the message again.

So the idea is, a subscriber should ack a message exactly when that message has been dealt with. A message has been dealt with when it has been successfully processed. But it also has been dealt with if the message is invalid, and the appropriate party has been notified of same. That is "failure" in one sense, but failing to ack that message is just going to give the same problem to someone else, so it's success in another sense. 

On the other hand, if the message is invalid and the subscribing process tries to send an email to the sender and the email system replies "down for maintenance, come back later," then the message shouldn't be acked, because it hasn't been dealt with. (The right thing to do here is probably to nack it, and also to tell the system to delay redelivery; but you might not want to get into that because some systems don't let you add a delay.)

So when I wrote 

> [Whether to ack] cross-cuts the question of whether processing the message "fails"

I meant that acks are about dealing with the message once and for all, and that may include cases you'd describe as failure.